add torch.autograd._unsafe_set_version_counter API

bdhirsh opened the pull request:
better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #93994
* #92857
* #94195
* #94018
* __->__ #92924

bdhirsh committed:
add torch.autograd._unsafe_decrement_version_counter API

[ghstack-poisoned]

ezyang commented:
Instead of manually doing a decrement, wouldn't it make more sense to just allow setting the version counter to some value? This would allow you to do multiple mutating operations and then reset it to some historical state.

bdhirsh commented:
that seems fair (and imo that also makes it clearer that this is an unsafe API). I'll update

bdhirsh committed:
Update on "add torch.autograd._unsafe_decrement_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

alanwaketan commented on c10/core/TensorImpl.h:
```diff
@@ -388,6 +388,15 @@ struct C10_API VariableVersion {
+    TORCH_CHECK(
+        version_counter_,
+        "Tried to call torch.autograd._unsafe_set_version() on a tensor "
+        "that does not have a version counter. Was it created it in inference mode?");
```
Nit: Was it created in inference mode?

alanwaketan reviewed (approved):
Thanks, Brian. Let me verify if it fixes the issue.

albanD reviewed (commented):
I still think that a "decrement" is better than a "set". It give a lot less room for the user to get things wrong. Could you share why you made it a set here?

bdhirsh commented:
Great! 

> Detailed error:

@alanwaketan just to confirm, are you saying that you're seeing an error? Or everything worked as expected

bdhirsh commented:
One other thing - @ezyang are you ok with a context manager, like `with torch.autograd._preserve_version_counter(tensor)`, that internally calls this `_unsafe_set_version_counter` API? I was talking to Alban, and he thinks this is slightly less of a footgun: we don't want people doing stuff like this in the future:
```
old_version = t._version
function_that_does_arbitrary_things_to_t(t)
# We might be ignoring legitimate mutations to t that autograd should have been aware of!
torch.autograd._unsafe_set_version_counter(old_version)
```

alanwaketan commented:
> Great!
> 
> > Detailed error:
> 
> @alanwaketan just to confirm, are you saying that you're seeing an error? Or everything worked as expected

Updated the confusing comment. That was the original error.

ezyang commented:
I am fine with context manager

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

albanD commented on torch/autograd/grad_mode.py:
```diff
@@ -254,6 +255,35 @@ def __exit__(self, *args) -> None:
         return self.__class__(self.mode)
 
+class _unsafe_preserve_version_counter(_DecoratorContextManager):
+    r"""Context-manager that unsafely tells autograd to ignore mutations to a tensor.
```
Could you make this more scarry please?
```suggestion
    r"""DO NOT USE THIS UNLESS YOU KNOW EXACTLY WHAT YOU'RE DOING!
    This context manager can lead to arbitrary silent-correctness issues in any other part of your code (even the ones not touched directly by the context manager)!
    
    Context-manager that unsafely tells autograd to ignore mutations to a tensor.
```

bdhirsh replied:
absolutely 😛

albanD reviewed (approved):
Just update to the doc. Sounds good otherwise.

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

alanwaketan commented on torch/autograd/grad_mode.py:
```diff
@@ -254,6 +255,35 @@ def __exit__(self, *args) -> None:
     def clone(self):
         return self.__class__(self.mode)
 
+class _unsafe_preserve_version_counter(_DecoratorContextManager):
```
Yea, it looks like it would work. I can give it a try as well. Will respond back.

albanD reviewed (approved):
Thanks!

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge -f "existing failure is on trunk"

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]

bdhirsh committed:
Update on "add torch.autograd._unsafe_set_version_counter API"

better description coming soon (but this is meant to fix https://github.com/pytorch/pytorch/issues/91093)




[ghstack-poisoned]
